### DISCUSSION TITLE
Query to find documents that don't have its Id in a different collection

### DISCUSSION THREAD

**Discussion by rallman** [ravendb/ravendb#16258] [Q&A]
We have an export service that allows for multiple external destinations for each exportable event. To handle the status of each destination (channel), we create a document for each combination of event and channel. Like this:

ExportChannel:
```
{
  "Id": "ExportChannel/1-A",
  "startDate": "2023-03-27T21:30:20.5489648"
  [...]
},
{
  "Id": "ExportChannel/2-A",
  "startDate": "2023-03-28T21:30:20.5489648"
  [...]
}
```

ExportEvent:
```
{
  "Id": "ExportEvent/1-A",
  "eventDate": "2023-04-03T21:30:20.5489648"
  [...]
}
```

ChannelEventStatus:
```
{
  "Id": "ChannelEventStatus/1-A".
  "EventId": "ExportEvent/1-A",
  "Complete": true
}
```

Our internal eventing does a pretty good job of creating the `ChannelEventStatus` documents but sometimes they don't get created. The channels are also created and the `startDate` is set in the past - and therefore historical events are expected to be sent. For both reasons, I would like to get a list of `ExportEvent` documents that do not have a corresponding `ChannelEventState` for a given `ExportChannel`.

In the above example, the document `ChannelEventStatus/1-A` would be returned when querying for `ExportChannel/2-A`.

Can this be accomplished in a single query of the `ChannelEventStatus` collection?
(right now, I'm more interested in the query capability than approaching the result using a different mechanism)

**Comment by ayende** [comment]
I'm not sure that I'm following.
You are correlating between the events and the channels based on what field?

You want all the channels that *do not* have a channel status document?

You can do that using a map/reduce index, where you are aggregating both the channels and the statuses and reducing by the id.
Then you can detect all where the number of channels & statuses doesn't both equal to 1.

Something like:

```
//map
from c in docs.Channels
select new 
{
   ChannelId = c.Id,
   ChannelCount = 1,
   StatusesCount = 0
}
// map
from cs in docs.ChannelStatus
select new
{
   cs.ChannelId,
   ChannelCount = 0,
   StatusesCount = 1
}

// reduce
from result in results
group result by result.ChannelId into g
select new
{
   ChannelId = g.Key,
   ChannelCount = g.Sum(x=>x.ChannelCount )
   StatusesCount = g.Sum(x=>x.StatusesCount)
}
```

And then query then becomes:

```
from index ChannelsAndStatuses
where StatusesCount == 0 or ChannelCount == 0
```

**Reply by rallman** [reply]
Sorry, I missed the related ChannelId on the ChannelEventState document. It should have looked like this:
```
{
  "Id": "ChannelEventStatus/1-A".
  "EventId": "ExportEvent/1-A",
  "ChannelId": "ExportChannel/1-A",
  "Complete": true
}
```

I'm trying to find all of the `ExportEvent` documents that do not have a corresponding `ChannelEventStatus` for a given `ExportChannel`.

It almost seems like it needs a fanout type index on `ExportEvent` with the list of `ExportChannels` providing the "many" side.

**Reply by garayx** [reply]
I think this can be done by slightly modifying the index above,


```
//map
from c in docs.ExportEvent
select new 
{
   EventId = c.Id,
   ExportEventCount = 1,
   ChannelEventCount = 0
}
// map
from cs in docs.ChannelEventStatus
select new
{
   EventId =cs.EventId,
   ExportEventCount = 0,
   ChannelEventCount = 1
}

// reduce
from result in results
group result by result.EventId into g
select new
{
   EventId = g.Key,
   ExportEventCount = g.Sum(x=>x.ExportEventCount )
   ChannelEventCount = g.Sum(x=>x.ChannelEventCount)
}
```

Query:

```
from index ChannelsAndStatuses
where ChannelEventCount == 0 
```
